### PR TITLE
chore: convert two slack channel ID char fields to foreign keys

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

Similar to https://github.com/grafana/oncall/pull/5199

Converts follow char fields to primary key relationships on `SlackChannel` table:
- `ResolutionNoteSlackMessage.channel_id` -> `ResolutionNoteSlackMessage.slack_channel`
- `ChannelFilter.slack_channel_id` -> `ChannelFilter.slack_channel`

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] Added the relevant release notes label (see labels prefixed w/ `release:`). These labels dictate how your PR will
    show up in the autogenerated release notes.

### CODE REVIEW

File: engine/apps/api/serializers/channel_filter.py [hunk 1]
```diff
@@ -232,5 +196,5 @@ class ChannelFilterUpdateResponseSerializer(ChannelFilterUpdateSerializer):
 
     slack_channel = serializers.SerializerMethodField()
 
-    def get_slack_channel(self, obj) -> SlackChannelDetails | None:
-        return self._get_slack_channel(obj)
+    def _get_slack_channel(self, obj) -> SlackChannelDetails | None:
+        pass
```
Comment (joeyorlando) [inline engine/apps/api/serializers/channel_filter.py:200]: this is only used in openapi schema "hints", I don't think this function _actually_ needs to do anything (this was the last reference to `_get_slack_channel`)

File: engine/apps/alerts/incident_log_builder/incident_log_builder.py
Comment (joeyorlando) [inline engine/apps/alerts/incident_log_builder/incident_log_builder.py:1]: I started poking around in here (for some reason?) and decided to leave the typing in a better spot than I found it.

The only changes in this file are _really_ related to improving typing, there's no functional changes

File: engine/apps/alerts/incident_log_builder/incident_log_builder.py [hunk 2]
```diff
@@ -120,25 +163,22 @@ def _get_resolution_notes(self) -> "RelatedManager['ResolutionNote']":
...
-        incident_escalation_plan = dict()
-        incident_escalation_plan = self._add_invitation_plan(incident_escalation_plan, for_slack=for_slack)
+    def get_escalation_plan(self, for_slack: bool = False) -> FlattendEscalationPlan:
+        escalation_plan: EscalationPlan = dict()
+        escalation_plan = self._add_invitation_plan(escalation_plan, for_slack=for_slack)
+
         if not self.alert_group.acknowledged and not self.alert_group.is_silenced_forever:
-            incident_escalation_plan = self._add_escalation_plan(incident_escalation_plan, for_slack=for_slack)
-        final_incident_escalation_plan = self._finalize_escalation_plan_dict(incident_escalation_plan)
-        return final_incident_escalation_plan
```
Comment (joeyorlando) [inline engine/apps/alerts/incident_log_builder/incident_log_builder.py:134]: saw an opportunity to (continue) removing references to "incident" (I found this is confusing wrt an "incident" from Grafana Incident).. updated references to this function

File: engine/apps/alerts/incident_log_builder/incident_log_builder.py [hunk 3]
```diff
@@ -10,26 +10,69 @@
...
+
+
 class IncidentLogBuilder:
     def __init__(self, alert_group: "AlertGroup"):
         self.alert_group = alert_group
 
-    def get_log_records_list(
-        self, with_resolution_notes: bool = False
-    ) -> typing.List[typing.Union["AlertGroupLogRecord", "ResolutionNote", "UserNotificationPolicyLogRecord"]]:
+    def get_log_records_list(self, with_resolution_notes: bool = False) -> LogRecords:
```
Comment (joeyorlando) [inline engine/apps/alerts/incident_log_builder/incident_log_builder.py:70]: only references to `IncidentLogBuilder` in `grafana/oncall-private` are to this function (nothing has changed here)

File: engine/apps/alerts/escalation_snapshot/snapshot_classes/escalation_policy_snapshot.py [hunk 4]
```diff
@@ -57,24 +60,24 @@ class EscalationPolicySnapshot:
...
+        num_alerts_in_window: typing.Optional[int],
+        num_minutes_in_window: typing.Optional[int],
+        custom_webhook: typing.Optional["Webhook"],
+        notify_schedule: typing.Optional["OnCallSchedule"],
+        notify_to_group: typing.Optional["SlackUserGroup"],
+        escalation_counter: int,
+        passed_last_time: typing.Optional[datetime.datetime],
+        pause_escalation: bool,
+        notify_to_team_members: typing.Optional["Team"] = None,
+        severity: typing.Optional[str] = None,
```
Comment (vadimkerr) [inline engine/apps/alerts/escalation_snapshot/snapshot_classes/escalation_policy_snapshot.py:80]: 👍

File: engine/apps/api/serializers/channel_filter.py (hunk 1 above)
Comment (vadimkerr) [reply engine/apps/api/serializers/channel_filter.py:202]: I _think_ we can remove this? `slack_channel` seems to be defined on the parent class as `SlackChannelSerializer` so drf-spectacular should be able to introspect OpenAPI types automatically?

**Review by vadimkerr — COMMENTED** [review]
LGTM. This is a great change, having proper FKs is really nice!

**Review by vadimkerr — APPROVED** [review]

File: engine/apps/api/serializers/channel_filter.py (hunk 1 above)
Comment (joeyorlando) [reply engine/apps/api/serializers/channel_filter.py:202]: good call, thanks for pointing this out! Just removed `ChannelFilterUpdateResponseSerializer` and double-checked swagger spec for `/channel_filters`, looks good 👍 
Screenshot 2024-11-04 at 1 13 09 PM
Screenshot 2024-11-04 at 1 13 31 PM

**Pull request merged**